### PR TITLE
[CI] Generate valid testdata that can be used for integration testing

### LINKED ISSUE #312: [CI] Testdata catalogs are invalid
# What happened?

On a connected cluster:
```shell
$ oc get pod -A -l app=packageserver
NAMESPACE                              NAME                             READY   STATUS    RESTARTS   AGE
openshift-operator-lifecycle-manager   packageserver-76c948d469-68gmw   1/1     Running   0          28m

$ oc apply -f- << EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: oc-mirror-dev
  namespace: openshift-marketplace
spec:
  displayName: oc-mirror Test Catalog
  image: quay.io/redhatgov/oc-mirror-dev:test-catalog-latest
  publisher: RH4G
  sourceType: grpc
EOF
catalogsource.operators.coreos.com/oc-mirror-dev created

$ oc get pod -A -l app=packageserver
NAMESPACE                              NAME                             READY   STATUS    RESTARTS   AGE
openshift-operator-lifecycle-manager   packageserver-76c948d469-68gmw   1/1     Running   0          29m

$ oc delete pod -l app=packageserver -n openshift-operator-lifecycle-manager
pod "packageserver-76c948d469-68gmw" deleted

$ oc get pod -A -l app=packageserver
NAMESPACE                              NAME                             READY   STATUS             RESTARTS      AGE
openshift-operator-lifecycle-manager   packageserver-76c948d469-q8pj4   0/1     CrashLoopBackOff   5 (15s ago)   3m29s

$ oc logs -n openshift-operator-lifecycle-manager -l app=packageserver
	/build/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators/packagemanifest.go:126 +0x19f
github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators.CreateCSVDescription(0xc00176e000, 0xc000f82000, 0x1b5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/build/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators/packagemanifest.go:38 +0x1e6
github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider.newPackageManifest(0x246e2c0, 0xc00208b440, 0xc002f5e8c0, 0xc001f11e00, 0xc000bde860, 0x0, 0x0, 0xc001f11e00)
	/build/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider/registry.go:506 +0xd14
github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider.(*RegistryProvider).refreshCache.func1(0xc001315090, 0xc000bde860, 0x246e2c0, 0xc00208b440, 0xc00150e4c0, 0xc002f5e8c0, 0xc000440ee0, 0xc001315088, 0xc0007e8450)
	/build/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider/registry.go:309 +0x2e5
created by github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider.(*RegistryProvider).refreshCache
	/build/vendor/github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/provider/registry.go:301 +0x57e
```

# What you expected to happen?

I expected the packageserver to be able to parse the catalog and provide a list of PackageManifests for listing in OperatorHub.

# How to reproduce it (as minimally and precisely as possible)?

With a connected OpenShift or CRC cluster:

```console
$ oc apply -f- << EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: oc-mirror-dev
  namespace: openshift-marketplace
spec:
  displayName: oc-mirror Test Catalog
  image: quay.io/redhatgov/oc-mirror-dev:test-catalog-latest
  publisher: RH4G
  sourceType: grpc
EOF
```

# Anything else we need to know?

It appears to be relatedImages that's causing the crash:
https://github.com/operator-framework/operator-lifecycle-manager/blob/aa20d1cea93e3e0a82aab1e586c28f935ffdd297/pkg/package-server/apis/operators/packagemanifest.go#L38

The relatedImages from the CSVs appear to be here:
https://github.com/openshift/oc-mirror/blob/de53a7e4fe7683b37836cafb5d31a9cb24f09574/test/operator/testdata/bundles/bar-bundle-v0.1.0/manifests/bar.csv.yaml#L16

And those appear to be getting manipulated here:
https://github.com/openshift/oc-mirror/blob/de53a7e4fe7683b37836cafb5d31a9cb24f09574/test/operator/setup-testdata.sh#L43

Except that I believe the result is, for example `quay.io/redhatgov/oc-mirror-dev/bar-operator/bar:v0.1.0` when it should be `quay.io/redhatgov/oc-mirror-dev:bar-v0.1.0`

### LINKED ISSUE #347: [CI] testdata bundles with deployable content

### PR DESCRIPTION
**Pull request by jharmison-redhat**
# Description

Test data catalog generation has been changed. The catalog is still rendered as a File-Based Catalog, however there is minimal templating of indices and instead a Python script that emulates some OPM behavior is used to parse the operator bundles and generate an index using bundle data and metadata.

Importantly, this means that changes to bundles and their impact on the generated catalogs can be more easily understood without changes in two areas at once. The catalogs can actually be used on an OpenShift cluster which means that proper end-to-end testing on a live OpenShift cluster can test behavior with OLM from known-state catalogs. We can control the channel graph of the indexes, mirror them to a mock disconnected environment, and validate that behavior aligns with what we expect.

There were a few other changes made to support this:
- `test/operator/testdata/bundles` moved to `test/operator/bundles`
- `test/operator/testdata/indices` removed entirely
- `test/operator/setup-testdata.sh` functionality to build bundles, indices, and related images (which was all turned off by a flag set in the script) moved to `test/operator/publish_images.py`, which also renders indices directly from the bundles
- `test/operator/testdata/configs/latest` moved to `test/configs`
- the things being used for e2e-simple in `test/operator/setup-testdata.sh` moved to `test/lib.sh`, references to `configs` adjusted, `e2e-simple.sh` updated to use new `lib.sh` functions.
- `test/prow` and `test/e2e-prow.sh` removed entirely

When the `make publish-catalog` target is run from the root of the repository, the [described list of bundles for each specified catalog](https://github.com/jharmison-redhat/oc-mirror/blob/b454991db749217cd5eccd57a492dd24227ce306/test/operator/publish_images.py#L25) are read and organized into descriptive Python objects that model how OPM thinks about the bundles. That is, bundles are put into channels for packages arranged in an index. When an index is rendered, the package, channel, and bundle entries are organized into a file-based catalog using metadata gleaned from the bundles. This includes upgrade graph data from bundle annotations, and the channels that are rendered represent the data from the bundles in that channel. After rendering the FBC, container images are built and pushed for the index, bundles, and a common "related_image" for each operator. This related_image is a simple distroless image with a busybox shell script which echoes its own published name and then sleeps indefinitely. Each operator bundle specifies the deployment of the operator using this image.

Note that right now, image references have _not_ been converted to manifest hash references. Actual disconnected testing has not been performed, but it is expected to be problematic because of that.

Fixes #347
Fixes #312

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

```
$ REGISTRY=registry.jharmison.com make
Rendering and publishing images
INFO: Rendering test-catalog-minimal catalog indexes.
INFO:  Processing foo index
INFO:   Rendering package foo
INFO:   Rendering channel beta
INFO:   Rendering bundle foo.v0.1.0
INFO:  Processing bar index
INFO:   Rendering package bar
INFO:   Rendering channel alpha
INFO:   Rendering bundle bar.v0.1.0
INFO: Building images for test-catalog-minimal catalog.
INFO:  Building and pushing registry.jharmison.com/redhatgov/oc-mirror-dev:test-catalog-minimal
INFO:  Building and pushing registry.jharmison.com/redhatgov/oc-mirror-dev:foo-bundle-v0.1.0
INFO:  Building and pushing registry.jharmison.com/redhatgov/oc-mirror-dev:foo-v0.1.0
INFO:  Building and pushing registry.jharmison.com/redhatgov/oc-mirror-dev:bar-bundle-v0.1.0
INFO:  Building and pushing registry.jharmison.com/redhatgov/oc-mirror-dev:bar-v0.1.0
---
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: oc-mirror-dev-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: registry.jharmison.com/redhatgov/oc-mirror-dev:test-catalog-minimal
  displayName: oc-mirror Development Testing Catalog
  publisher: RedHatGov
  updateStrategy:
    registryPoll:
      interval: 30m
INFO: Cleaning up /home/james/Projects/oc-mirror/test/operator/operator-build-minimal-xxm5axgb
INFO: Rendering test-catalog-minimal-diff catalog indexes.
[...SNIP...]
```

```
$ oc new-project bar-operator
Now using project "bar-operator" on server "https://api.crc.testing:6443".
You can add applications to this project with the 'new-app' command. For example, try:
    oc new-app rails-postgresql-example
to build a new example application in Ruby. Or use kubectl to deploy a simple Kubernetes application:
    kubectl create deployment hello-node --image=k8s.gcr.io/e2e-test-images/agnhost:2.33 -- /agnhost serve-hostname
```

```
$ oc get catalogsource -n openshift-marketplace
NAME                  DISPLAY               TYPE   PUBLISHER   AGE
certified-operators   Certified Operators   grpc   Red Hat     21d
community-operators   Community Operators   grpc   Red Hat     21d
redhat-marketplace    Red Hat Marketplace   grpc   Red Hat     21d
redhat-operators      Red Hat Operators     grpc   Red Hat     21d
```

```
$ oc apply -f- <<EOF
---
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: oc-mirror-dev-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: registry.jharmison.com/redhatgov/oc-mirror-dev:test-catalog-latest
  displayName: oc-mirror Development Testing Catalog
  publisher: RedHatGov
  updateStrategy:
    registryPoll:
      interval: 30m
EOF
catalogsource.operators.coreos.com/oc-mirror-dev-catalog created
```

```
$ oc operator list-available | grep bar
bar                                                oc-mirror Development Testing Catalog  alpha                               bar.v0.2.0                                               15s
bar                                                oc-mirror Development Testing Catalog  stable                              bar.v1.0.0                                               15s
```

```
$ oc operator install -a Automatic -c alpha bar -C
operatorgroup "bar-operator" created
subscription "bar" created
operator "bar" installed; installed csv is "bar.v0.2.0"
```

```
$ oc logs deploy/bar-operator
relatedImage: registry.jharmison.com/redhatgov/oc-mirror-dev:bar-v0.2.0
```

```
$ oc new-project foo-operator
Now using project "foo-operator" on server "https://api.crc.testing:6443".
You can add applications to this project with the 'new-app' command. For example, try:
    oc new-app rails-postgresql-example
to build a new example application in Ruby. Or use kubectl to deploy a simple Kubernetes application:
```

```
$ oc operator install -a Automatic -c beta foo -C
operatorgroup "foo-operator" created
subscription "foo" created
operator "foo" installed; installed csv is "foo.v0.3.1"
```

```
$ oc logs deploy/foo-operator
relatedImage: registry.jharmison.com/redhatgov/oc-mirror-dev:foo-v0.3.1
```

```
$ oc operator list
PACKAGE  SUBSCRIPTION                                           INSTALLED CSV  CURRENT CSV  STATUS         AGE
bar      bar                                                    bar.v0.2.0     bar.v0.2.0   AtLatestKnown  52s
bar      bar-alpha-oc-mirror-dev-catalog-openshift-marketplace  bar.v0.2.0     bar.v0.2.0   AtLatestKnown  12s
foo      foo                                                    foo.v0.3.1     foo.v0.3.1   AtLatestKnown  18s
```

```
$ oc patch catalogsource oc-mirror-dev-catalog -n openshift-marketplace --type=merge -p '{"spec":{"image":"registry.jharmison.com/redhatgov/oc-mirror-dev:test-catalog-diff"}}'
catalogsource.operators.coreos.com/oc-mirror-dev-catalog patched
```

```
$ oc delete pod -n openshift-marketplace $(oc get pod -n openshift-marketplace | awk '/oc-mirror/{print $1}')
pod "oc-mirror-dev-catalog-x74fm" deleted
```

```
$ oc get pod -n openshift-marketplace -w
NAME                                                              READY   STATUS      RESTARTS   AGE
0af2e4ba91a71acbc60641281453c447fb0295df10f7c3334590762fb15x82r   0/1     Completed   0          62s
289956107b7ed86017e7a14a81e6cdfa9b584577dc5344adbfc142a018gvrt9   0/1     Completed   0          96s
certified-operators-4bxzn                                         1/1     Running     0          21d
community-operators-kvkdn                                         1/1     Running     0          21d
marketplace-operator-7457f7588f-tn8zd                             1/1     Running     0          21d
oc-mirror-dev-catalog-s5nrm                                       0/1     Running     0          6s
redhat-marketplace-xhg2d                                          1/1     Running     0          21d
redhat-operators-nbctt                                            1/1     Running     0          20d
oc-mirror-dev-catalog-s5nrm                                       1/1     Running     0          10s
```

```
$ oc operator list-available | grep foo
foo                                                oc-mirror Development Testing Catalog  stable                              foo.v0.2.0                                               2m18s
foo                                                oc-mirror Development Testing Catalog  beta                                foo.v0.3.2                                               2m18s
```

```
$ oc operator list
PACKAGE  SUBSCRIPTION                                           INSTALLED CSV  CURRENT CSV  STATUS         AGE
bar      bar                                                    bar.v0.2.0     bar.v0.2.0   AtLatestKnown  118s
bar      bar-alpha-oc-mirror-dev-catalog-openshift-marketplace  bar.v0.2.0     bar.v0.2.0   AtLatestKnown  78s
foo      foo                                                    foo.v0.3.2     foo.v0.3.2   AtLatestKnown  84s
```

```
$ oc logs deploy/foo-operator
relatedImage: registry.jharmison.com/redhatgov/oc-mirror-dev:foo-v0.3.2
```

Additionally, to validate that the changes to `test/lib.sh` and `test/e2e-simple.sh` worked as expected, ran an additional `make test-e2e` from the root Makefile.

**Test Configuration**:
Fedora 35
OpenShift 4.10.3
podman version 3.4.4
python 3.10.3

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jharmison-redhat** [comment]
@jpower432 afaict this is the test failing with the testdata that's published to Quay, which I have not replaced. Can you confirm the e2e-simple design up to this point?

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oc-mirror/389/pull-ci-openshift-oc-mirror-main-e2e/1508815376643788800#1:build-log.txt%3A6081

My local `make test-e2e` succeeded. I thought it may have been because I had exported `REGISTRY` and pointed to my images, but now I am less sure.

**Comment by jharmison-redhat** [comment]
I reran my local test as well and it's failing now, too. I'll need to figure out why and fix that.

**Comment by jpower432** [comment]
> @jpower432 afaict this is the test failing with the testdata that's published to Quay, which I have not replaced. Can you confirm the e2e-simple design up to this point?
> 
> https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oc-mirror/389/pull-ci-openshift-oc-mirror-main-e2e/1508815376643788800#1:build-log.txt%3A6081
> 
> My local `make test-e2e` succeeded. I thought it may have been because I had exported `REGISTRY` and pointed to my images, but now I am less sure.

@jharmison-redhat This failure you are seeing is a known issue with `opm diff`. I have a PR to fix the problem, but this is not a problem introduce by your changes.

**Comment by jharmison-redhat** [comment]
I can confirm that the deployments are created with tag references still, so the catalog generation will need to translate those to manifest hash references before writing the indexes.

**Comment by jharmison-redhat** [comment]
Ready for review, @jpower432

File: docs/dev/test-indexes.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+Environment variables or command line flags can override the registry, registry repository (or catalog namespace), container runtime, catalog selection, file-based catalog architecture, and apply a catalog to a cluster automatically. The command-line flags are best documented by the help screen for the script, visible at `publish_images.py -h`. Images should only be published to the default registry and catalog namespace from `main` to ensure that the main tests are drawing from known catalogs. If you want to publish alternative catalogs/bundles with local changes to test something before merging to main, the easiest way is to use either the `REGISTRY` or `CATALOG_NAMESPACE` variable to override the defaults. For example:
+
+```sh
+REGISTRY=registry.jharmison.com ./publish_images.py
```
Comment (jpower432) [inline docs/dev/test-indexes.md:118]: Are we going to change the reference here to `quay.io`?

Comment (jharmison-redhat) [reply docs/dev/test-indexes.md:118]: That's the default, this is to demonstrate how to override the registry references in every bundle and push to an alternative registry.

**Review by jpower432 — APPROVED** [review]
Testing the `publish-catalogs` functionality worked as expected. I was able to publish a new catalog to a local registry and mirror it with `oc-mirror` without issue
Nit: It would be nice if there was a `use-http` feature so that we could publish to local non-TLS registered, but I understand that this is not 100% intended for that.

**Pull request merged**